bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/VariableClearing.scala

```diff
@@ -0,0 +1,312 @@
+      } :+ withClearsMap(undecided) {
+        walkExpr(elseExpression)
+        clearVal(elseExpression)
+        BranchClearsMap(this.clearsMap, ClearSite(elseExpression, true))
+      }): _*)
+      
+    case _ => super.walkExpr(expression)
+  }
+  
+  var nodeSubst: IdentityHashMap[StatOrExpr, StatOrExpr] = new IdentityHashMap()
+  var nodeClears: IdentityHashMap[StatOrExpr, Buffer[Clear]] = new IdentityHashMap()
+  def withNodeClears[A](nodeReplacements: IdentityHashMap[StatOrExpr, Buffer[Clear]])(f: => A): A = {
+    val oldNodeClears = this.nodeClears
+    this.nodeClears = nodeReplacements
+    try {
+      return f
+    } finally {
+      this.nodeClears = oldNodeClears
+    }
+  }
+  
+  def clearsMap2nodeClears(clearsMap: ClearsMap): IdentityHashMap[StatOrExpr, Buffer[Clear]] = {
+    val nodeClears = new IdentityHashMap[StatOrExpr, Buffer[Clear]]()
+    for ((sym, clears) <- clearsMap if clears != null) { // TODO WHAT IF UNENCOUNTERED?
+      for (ClearSite(node, before) <- clears) {
+        var buffer = nodeClears.get(node)
+        if (buffer == null) {
+          buffer = ArrayBuffer[Clear]()
+          nodeClears.put(node, buffer)
+        }
+        buffer += Clear(sym, before)
+      }
+    }
+    
+    nodeClears
+  }
+  
+  def replaceStat(statement: Statement): Statement = {
+    val replacements = nodeClears.get(statement)
+    if (replacements != null) {
+      treeCopy.ClearVarsStatement(statement, super.transformStat(statement),
+          replacements.filter(_.before).map(_.sym),
+          replacements.filter(!_.before).map(_.sym))
+    } else {
+  	  super.transformStat(statement)
+    }
+  }
+  
+  def replaceExpr(expression: Expression): Expression = {
+    val replacements = nodeClears.get(expression)
+      if (replacements != null) {
+        treeCopy.ClearVarsExpression(expression, super.transformExpr(expression),
+          replacements.filter(_.before).map(_.sym),
+          replacements.filter(!_.before).map(_.sym))
+      } else {
+    	  super.transformExpr(expression)
+      }
+  }
+  
+  override def transformExpr(expression: Expression) = expression match {
+    case proc @ ProcExpression(name, args, body, flags) =>
+      val clearsMap = withClearsMap(new HashMap()) {
+        args.foreach (_ match { case Variable(sym) => this.clearsMap.put(sym, null) })
+    	  walkStat(body)
+    	  this.clearsMap
+      }
+      withNodeClears(clearsMap2nodeClears(clearsMap)) {
+    	  treeCopy.ProcExpression(expression, name, args, replaceStat(body), flags)
+      }
+      
+    case v @ Variable(sym) => // Some nodes expect a variable as child. Do not create unnecessary resetters
+      val repl = nodeClears.get(v)
+      if (repl != null) {
+        val idx = repl.indexWhere { clear => clear.sym == sym }
+        if (idx >= 0) {
+          v.clear = true
+          if (repl.length == 1) {
+            nodeClears.put(v, null)
+          } else {
+            repl.remove(idx)
+          }
+        }
+      }
```

Comment:
This should not be needed if `(Variable)` casts in the Translator are replaced by a helper which can see through ClearVarsExpression.